[Webhook] re-structure and fill out missing parts (#21832)
symfony/symfony-docs: webhook.rst

    The Webhook component was introduced in Symfony 6.3.

> Webhooks are event notification mechanisms typically delivered via HTTP POST requests, enabling real-time updates between systems.

Reviewer: @ckrack what do you think about this rephrase?

```suggestion
In web-development a Webhook is a remote-notification mechanism, typically delivered via
HTTP POST requests.
Webhooks enable real-time updates between systems.
```

Author: Yes, sounds great. thank you

The Webhook component provides two primary capabilities:

1. **Consuming**: Receive and process webhook calls from remote systems

> 2. **Sending**: Dispatch webhooks to registered endpoints when events occur

Reviewer: ```suggestion
1. **Consuming**: Receive and process webhook calls from remote systems;
2. **Sending**: Dispatch webhooks callbacks to registered endpoints when events occur.
```

Author: thank you, I've updated it but left webhook as singular when followed by the plural callbacks.

The easiest way is using the maker command:

.. code-block:: terminal

    $ php bin/console make:webhook

.. tip::

    Starting in `MakerBundle`_ ``v1.58.0``, the ``make:webhook`` command generates both the parser and consumer classes, and updates your configuration automatically.

When extending :class:`Symfony\\Component\\Webhook\\Client\\AbstractRequestParser`, you need to implement two methods:

**getRequestMatcher()** - Validate the incoming request format:

::

    // src/Webhook/AcmeWebhookRequestParser.php
    namespace App\Webhook;

    use Symfony\Component\HttpFoundation\Request;
    use Symfony\Component\HttpFoundation\RequestMatcher\ChainRequestMatcher;
    use Symfony\Component\HttpFoundation\RequestMatcher\IsJsonRequestMatcher;
    use Symfony\Component\HttpFoundation\RequestMatcher\MethodRequestMatcher;
    use Symfony\Component\HttpFoundation\RequestMatcher\RequestMatcherInterface;
    use Symfony\Component\RemoteEvent\RemoteEvent;
    use Symfony\Component\Webhook\Client\AbstractRequestParser;

>     final class AcmeWebhookRequestParser extends AbstractRequestParser

Author: Should I also include a part about testing the custom parser? There's a helper people might not be aware of.

                406,
                $e->getMessage(),
                $e
            );
        }
    }

Consuming the RemoteEvent
~~~~~~~~~~~~~~~~~~~~~~~~~

Whether processed synchronously or asynchronously (via Messenger), you
need a consumer implementing
:class:`Symfony\\Component\\RemoteEvent\\Consumer\\ConsumerInterface`.

The ``make:webhook`` command generates one automatically. Otherwise, create
it manually using the
:class:`Symfony\\Component\\RemoteEvent\\Attribute\\AsRemoteEventConsumer`
attribute:

::

    // src/RemoteEvent/AcmeWebhookConsumer.php
    namespace App\RemoteEvent;

    use Symfony\Component\RemoteEvent\Attribute\AsRemoteEventConsumer;
    use Symfony\Component\RemoteEvent\Consumer\ConsumerInterface;
    use Symfony\Component\RemoteEvent\RemoteEvent;

    #[AsRemoteEventConsumer('acme_webhook')]  // must match routing name
    final class AcmeWebhookConsumer implements ConsumerInterface
    {
        public function consume(RemoteEvent $event): void
        {
            // Handle the event based on your business logic
        }
    }

The routing name in the attribute must match the configuration entry of the
webhook routing.

> Asynchronous Processing

Reviewer: ```suggestion
Asynchronous Consuming
```

:class:`Symfony\\Component\\RemoteEvent\\Event\\Sms\\SmsEvent`:

::

    // src/RemoteEvent/SmsWebhookConsumer.php
    namespace App\RemoteEvent;

    use Symfony\Component\RemoteEvent\Attribute\AsRemoteEventConsumer;
    use Symfony\Component\RemoteEvent\Consumer\ConsumerInterface;
    use Symfony\Component\RemoteEvent\Event\Sms\SmsEvent;
    use Symfony\Component\RemoteEvent\RemoteEvent;

    #[AsRemoteEventConsumer('notifier_twilio')]
    final class SmsWebhookConsumer implements ConsumerInterface
    {
        public function consume(RemoteEvent $event): void
        {
            if ($event instanceof SmsEvent) {
                $this->handleSms($event);
            }
        }

        private function handleSms(SmsEvent $event): void
        {
            // Update SMS delivery status in database, etc.
        }
    }

Sending Webhooks
----------------

The Webhook component also enables your application to dispatch webhooks to
external endpoints when your application events occur. This is useful when
building APIs that notify subscribers of important events.

.. note::

    To send webhooks, ensure you have installed both the HttpClient and Serializer

>     component via ``composer require symfony/http-client symfony/serializer``.

Reviewer 2: serializer isn't really required, webhook component ships with a json_encode serializer and uses symfony's serializer only if it is available.

Author: That is true for 7.2 (introduced in https://github.com/symfony/symfony/pull/57881), but it was coupled to the serializer in 6.4, which we target here.
I'll go through the changelogs and mark the relevant changes, there's also the switch to php based routing.

Vonage       ``notifier.webhook.request_parser.vonage``
============ ==========================================

Configure similarly to mailers, then consume
:class:`Symfony\\Component\\RemoteEvent\\Event\\Sms\\SmsEvent`:

::

    // src/RemoteEvent/SmsWebhookConsumer.php
    namespace App\RemoteEvent;

    use Symfony\Component\RemoteEvent\Attribute\AsRemoteEventConsumer;
    use Symfony\Component\RemoteEvent\Consumer\ConsumerInterface;
    use Symfony\Component\RemoteEvent\Event\Sms\SmsEvent;
    use Symfony\Component\RemoteEvent\RemoteEvent;

    #[AsRemoteEventConsumer('notifier_twilio')]
    final class SmsWebhookConsumer implements ConsumerInterface
    {
        public function consume(RemoteEvent $event): void
        {
            if ($event instanceof SmsEvent) {
                $this->handleSms($event);
            }
        }

        private function handleSms(SmsEvent $event): void
        {
            // Update SMS delivery status in database, etc.
        }
    }

Sending Webhooks
----------------

The Webhook component also enables your application to dispatch webhooks to
external endpoints when your application events occur. This is useful when

> building APIs that notify subscribers of important events.

Reviewer: ```suggestion
The Webhook component also enables your application to invoke webhooks on
remote endpoints.
This is useful, as an example, when building APIs that notify subscribers of important events.
```

.. note::

    To send webhooks, ensure you have installed both the HttpClient and Serializer
    component via ``composer require symfony/http-client symfony/serializer``.

Basic Usage
~~~~~~~~~~~

To send a webhook, dispatch a
:class:`Symfony\\Component\\Webhook\\Messenger\\SendWebhookMessage` via
the Messenger component:

::

    use Symfony\Component\Messenger\MessageBusInterface;
    use Symfony\Component\RemoteEvent\RemoteEvent;
    use Symfony\Component\Webhook\Messenger\SendWebhookMessage;
    use Symfony\Component\Webhook\Subscriber;

    // In a service or controller

    $subscriber = new Subscriber(
        url: 'https://example.com/webhook/symfony',
        secret: 'your-shared-secret'
    );

    $event = new RemoteEvent(
        name: 'resource.created',
        id: '550e8400-e29b-41d4-a716-446655440000',
        payload: [
            'resource_id' => 12345,
            'email' => 'user@example.com',
            'created_at' => time(),
        ]
    );

    $this->messageBus->dispatch(
        new SendWebhookMessage($subscriber, $event)
    );

> The message will be queued and processed by

Reviewer: ```suggestion
The message will be queued and processed by the
```

:class:`Symfony\\Component\\Webhook\\Messenger\\SendWebhookHandler`,
which:

1. Constructs the HTTP request body (JSON-encoded payload)
2. Adds standard headers:

   - ``Webhook-Event``: The event name
   - ``Webhook-Id``: The event ID
   - ``Webhook-Signature``: HMAC-SHA256 signature (configurable) of the
     concatenated event name, ID, and body
   - ``Content-Type``: application/json

3. Signs the request using the subscriber's secret

> 4. Sends via Symfony's HttpClient component

Reviewer: ```suggestion
4. Sends the HTTP request by using the Symfony's HttpClient component
```

Author: Thanks, although the combination of "the" and `'` in `Symfony's` is incorrect grammar. I've removed the `'`.

Resulting HTTP Request
~~~~~~~~~~~~~~~~~~~~~~

When the webhook is sent, it generates an HTTP POST request with the following format:

.. code-block:: text

    POST /webhook/symfony HTTP/1.1
    Host: example.com
    Content-Type: application/json
    Webhook-Event: resource.created
    Webhook-Id: 550e8400-e29b-41d4-a716-446655440000
    Webhook-Signature: sha256=9f86d081884c7d6d9ffd60bb51d3263112c4b2486f80fa12ab5807265dc789d6

    {
        "resource_id": 12345,
        "email": "user@example.com",
        "created_at": 1234567890
    }

By default, the signature uses HMAC-SHA256 of the concatenated event name, event ID, and JSON body,
encoded in base64. Receiving endpoints should verify this signature using the shared secret
to ensure webhook authenticity.

Custom Sending Logic
~~~~~~~~~~~~~~~~~~~~

> For advanced use cases, you can build custom sending logic using

Reviewer: ```suggestion
For advanced use cases, you can build a custom sending logic by implementing the
```

Author: At first I wanted to go without the "a", but actually you can have only one custom logic - so that makes sense. Thanks.

::

    // src/RemoteEvent/SmsWebhookConsumer.php
    namespace App\RemoteEvent;

    use Symfony\Component\RemoteEvent\Attribute\AsRemoteEventConsumer;
    use Symfony\Component\RemoteEvent\Consumer\ConsumerInterface;
    use Symfony\Component\RemoteEvent\Event\Sms\SmsEvent;
    use Symfony\Component\RemoteEvent\RemoteEvent;

    #[AsRemoteEventConsumer('notifier_twilio')]
    final class SmsWebhookConsumer implements ConsumerInterface
    {
        public function consume(RemoteEvent $event): void
        {
            if ($event instanceof SmsEvent) {
                $this->handleSms($event);
            }
        }

        private function handleSms(SmsEvent $event): void
        {
            // Update SMS delivery status in database, etc.
        }
    }

Sending Webhooks
----------------

The Webhook component also enables your application to dispatch webhooks to
external endpoints when your application events occur. This is useful when
building APIs that notify subscribers of important events.

.. note::

    To send webhooks, ensure you have installed both the HttpClient and Serializer
    component via ``composer require symfony/http-client symfony/serializer``.

> .. versionadded:: 7.2

Author: > You are not allowed to use version "7.2". Only major version "6" is allowed.

welp. is this not allowed when targetting 6.4? But i can find other occurrences, so I'll live it here.

Reviewer 3: When targeting 6.4, the documentation only mention features which are available in this version. For something added in 7.x you'll need to create a new pull request against 7.4

Author: Thanks for the clarification. I've dropped the commit for now and moved it to a branch, that can be added after merging this.
